apis/mutations/unversioned/value.go

```diff
@@ -35,34 +36,51 @@ type AssignField struct {
 
 	// FromMetadata assigns a value from the specified metadata field.
 	FromMetadata *FromMetadata `json:"fromMetadata,omitempty"`
+
+	// ExternalData describes the external data source to be used for mutation.
+	ExternalData *ExternalData `json:"externalData,omitempty"`
 }
 
-func (a *AssignField) GetValue(metadata types.MetadataGetter) (interface{}, error) {
+func (a *AssignField) GetValue(param *types.GetValueParameter) (interface{}, error) {
 	if a == nil {
 		return nil, fmt.Errorf("assign is nil: %w", ErrInvalidAssignField)
 	}
-	if a.FromMetadata != nil {
-		return a.FromMetadata.GetValue(metadata)
+	if a.Value != nil {
+		return a.Value.GetValue(), nil
+	}
+	if a.ExternalData != nil {
+		return a.ExternalData.GetValue(param)
```

Comment:
We probably don't want to immediately reach out to the external data provider mid-mutation, for performance reasons.

K8s mutation webhooks are especially latency sensitive, given that they must be called serially. Also, Gatekeeper mutators can be called more than once on the same request (maximum of N calls, where N is the number of extant mutators).

Worst case scenario: we have N mutators that each call external data, the mutators are re-applied N times. (we can ignore that K8s uses mutation webhook reinvocation so our mutation webhook is invoked twice, since that is baked into our timeout). This leads to ~N^ 2 calls to a data provider. Let's assume N = 10. That would be 100 calls. With a timeout of 1 second, these providers would need to respond in <10ms to avoid a timeout.

IIRC the design doc had the idea of "placeholders", with the following execution flow:

- During mutation, assign a `placeholder` object to a field who has been mutated by a mutator leveraging external data. This `placeholder` object should include: the original value of the field, the external data provider to call.
- If another mutator mutates the same field, update the existing placeholder object to point to the new external data provider
- After the mutation process has completed, recurse through the finished object, gathering all `placeholder` objects
- Call all external data providers referenced by `placeholder` objects, with the following rules:
  - calls are done in parallel
  - calls to the same data provider are batched (i.e. only 1 request is sent out)
  - placeholder objects are replaced with the response from the data provider

With batching/parallel/lazy evaluation, the above scenario means that the worst-performing data provider should respond in ~500ms (this is assuming all the rest of the request handling takes ~500ms, which is pessimistic). It also means that irrelevant mutators (e.g. those that are overwritten by some other mutator) don't send unnecessary network requests.

Reply:
The part that mentions this in the DD starts with `# Lazy evaluation of provider promises (LEPP)` (though it's not super-obvious what the role of that placeholder object is)

Reply:
Thanks for the explanation. Let me revise the PR.

Reply:
Regarding the first rule above (calls are done in parallel), it means that we will need to restrict each field to be mutated by at most one external data provider. If we have two or more unique external data providers mutating the same field, processing them in parallel will mostly likely yield different values. Did I interrupt it correctly?